utils/include/detray/detectors/build_toy_detector.hpp

```diff
@@ -0,0 +1,804 @@
+        m_cyl_map_bins = {n_phi, n_z};
+        return *this;
+    }
+    constexpr toy_config &disc_map_bins(const std::size_t n_r,
+                                        const std::size_t n_phi) {
+        m_disc_map_bins = {n_r, n_phi};
+        return *this;
+    }
+    constexpr toy_config &thickness(const scalar_t t) {
+        assert(t > 0.f);
+        m_thickness = t;
+        return *this;
+    }
+    constexpr toy_config &mapped_material(const material<scalar_t> &mat) {
+        m_mapped_material = mat;
+        return *this;
+    }
+    constexpr toy_config &do_check(const bool check) {
+        m_do_check = check;
+        return *this;
+    }
+    /// @}
+
+    /// Getters
+    /// @{
+    constexpr unsigned int n_brl_layers() const { return m_n_brl_layers; }
+    constexpr unsigned int n_edc_layers() const { return m_n_edc_layers; }
+    constexpr scalar_t beampipe_vol_radius() const {
+        return m_beampipe_volume_radius;
+    }
+    constexpr bool use_material_maps() const { return m_use_material_maps; }
+    constexpr const std::array<std::size_t, 2> &cyl_map_bins() const {
+        return m_cyl_map_bins;
+    }
+    constexpr const std::array<std::size_t, 2> &disc_map_bins() const {
+        return m_disc_map_bins;
+    }
+    constexpr scalar_t thickness() const { return m_thickness; }
+    auto barrel_mat_generator() const { return m_cyl_mat_generator; }
+    auto edc_mat_generator() const { return m_disc_mat_generator; }
+    constexpr material<scalar_t> mapped_material() const {
+        return m_mapped_material;
+    }
+    constexpr const auto &barrel_layer_radii() const {
+        return m_barrel_layer_radii;
+    }
+    constexpr const auto &endcap_layer_positions() const {
+        return m_endcap_layer_positions;
+    }
+    constexpr const auto &outer_radius() const { return m_outer_radius; }
+    constexpr const auto &barrel_layer_binning() const {
+        return m_barrel_binning;
+    }
+    constexpr barrel_generator_config<scalar_t> &barrel_config() {
+        return m_barrel_factory_cfg;
+    }
+    constexpr endcap_generator_config<scalar_t> &endcap_config() {
+        return m_endcap_factory_cfg;
+    }
+    constexpr bool do_check() const { return m_do_check; }
+    /// @}
+};
+
+namespace detail {
+
+/// Add the cylinder and disc portals for a volume from explicit parameters
+///
+/// @param v_builder the volume builder to add the portals to
+/// @param names name map for volumes of the detector under construction
+/// @param vol_z half length of the cylinder
+/// @param h_z half length of the cylinder
+/// @param inner_r inner volume radius
+/// @param outer_r outer volume radius
+/// @param link_north portal volume link of the outer cylinder
+/// @param link_south portal volume link of the inner cylinder
+/// @param link_east portal volume link of the left disc
+/// @param link_west portal volume link of the right disc
+template <typename detector_t, typename scalar_t>
```

Comment:
You could spare the second (`scalar_t`) template argument on a lot of these functions. 🤔 The `detector_t` type already has the scalar type encoded in it. Meaning that you could quite easily generate code that would use `double` for `scalar_t`, while `detector_t` was set up with `float`. (Since the `scalar_t` type could be auto-deduced by the compiler from an innocent `1.`, or something similar.)

I agree that it's uglier, but you could use `typename detector_t::scalar_type` instead. 🤔 (In the function body you can create a typedef for it, but in the function signature you'd have to use this long-form name.)

Reply:
True, will change. The compiler would scream at me though, because of the following autoconversions